app/celery/nightly_tasks.py

```diff
@@ -73,45 +78,74 @@ def delete_notifications_older_than_retention():
 
 
 @notify_celery.task(name="delete-letter-notifications")
 @cronitor("delete-letter-notifications")
 def delete_letter_notifications_older_than_retention():
-    start = datetime.utcnow()
-    deleted = delete_notifications_older_than_retention_by_type('letter')
+    _delete_notifications_older_than_retention_by_type('letter')
+
+
+def _delete_notifications_older_than_retention_by_type(notification_type):
     current_app.logger.info(
-        "Delete {} job started {} finished {} deleted {} letter notifications".format(
-            'letter',
-            start,
-            datetime.utcnow(),
-            deleted
+        'Deleting {} notifications for services with flexible data retention'.format(notification_type))
+
+    flexible_data_retention = fetch_service_data_retention_for_all_services_by_notification_type(notification_type)
+    for f in flexible_data_retention:
+        day_to_delete_backwards_from = get_london_midnight_in_utc(
+            convert_utc_to_bst(datetime.utcnow()).date() - timedelta(days=f.days_of_retention)
         )
+
+        delete_notifications_for_service_and_type.apply_async(queue=QueueNames.REPORTING, kwargs={
+            'service_id': f.service_id,
+            'notification_type': notification_type,
+            'datetime_to_delete_before': day_to_delete_backwards_from
+        })
+
+    seven_days_ago = get_london_midnight_in_utc(convert_utc_to_bst(datetime.utcnow()).date() - timedelta(days=7))
+    service_ids_with_data_retention = {x.service_id for x in flexible_data_retention}
+
+    # get a list of all service ids that we'll need to delete for. Typically that might only be 5% of services.
+    # This query takes a couple of mins to run.
+    service_ids_that_have_sent_notifications_recently = get_service_ids_that_have_notifications_from_before_timestamp(
+        notification_type,
+        seven_days_ago
+    )
+
+    service_ids_to_purge = service_ids_that_have_sent_notifications_recently - service_ids_with_data_retention
+
+    current_app.logger.info('Deleting {} notifications for {} services without flexible data retention'.format(
+        notification_type,
+        len(service_ids_to_purge)
+    ))
+
+    for service_id in service_ids_to_purge:
+        delete_notifications_for_service_and_type.apply_async(queue=QueueNames.REPORTING, kwargs={
+            'service_id': service_id,
+            'notification_type': notification_type,
+            'datetime_to_delete_before': seven_days_ago
+        })
+
+
+@notify_celery.task(name='delete-notifications-for-service-and-type')
+def delete_notifications_for_service_and_type(service_id, notification_type, datetime_to_delete_before):
+    start = datetime.utcnow()
+    num_deleted = move_notifications_to_notification_history(
+        notification_type,
+        service_id,
+        datetime_to_delete_before,
+    )
+    end = datetime.utcnow()
+    current_app.logger.info(
+        f'Deleted {num_deleted} {notification_type} notifications for '
```

Comment:
Can we change this log message so that it's easier to search? For instance if the message starts with something like `delete-notifications-for-service-and-type has deleted {notification_type} for service {id} in {end-start} seconds`

Reply:
yes! i thought that the other day and then promptly forgot about it, that's a great idea.